### PR TITLE
If association is a hash-like object preloading fails

### PR DESCRIPTION
**Pull request by bpohoriletz**
### If association is a hash-like object preloading fails

If you pass a hash-like object to preload associations (for example ActionController::Parameters)
preloader will fail with the ArgumentError.
This change allows passing objects that may be converted to a Hash or String into a preloader

### Other Information

` bundle exec ruby -Itest test/cases/associations/eager_test.rb` passes

### CODE REVIEW

File: activerecord/lib/active_record/associations/preloader.rb
Comment (composerinteralia) [inline activerecord/lib/active_record/associations/preloader.rb]: I think we should get rid of the `true ==` and use `if/elsif`

File: activerecord/test/cases/associations/eager_test.rb
Comment (composerinteralia) [inline activerecord/test/cases/associations/eager_test.rb]: Does the specific `post_id` matter in these tests? Can we use `Post.preload(...).first` instead?

File: activerecord/test/cases/associations/eager_test.rb
Comment (composerinteralia) [inline activerecord/test/cases/associations/eager_test.rb]: Could avoid the passive voice here and write `"preloading works with object that responds to :to_hash"`

**Comment by bpohoriletz** [comment]
`bundle exec rake test:sqlite3` passes

File: activerecord/test/cases/associations/eager_test.rb
Comment (composerinteralia) [inline activerecord/test/cases/associations/eager_test.rb]: If this specific range is not important to the test, you can use `Object.new`. You could also just use a string or some other object that already responds to `to_sym`.

**Review by composerinteralia — CHANGES_REQUESTED** [review]
Thanks for the PR! This may be making the method more flexible than it needs to be, but  I suppose if an object is hash-like or string-like enough to respond to `to_hash` or `to_str` you might expect it to work here.

Could you also squash this into one commit?

~~Don't worry about the Travis failures on Ruby 2.5.0 `av:ujs` and `railties`. There is a Travis bug that has caused them to fail for a while.~~ Should be fixed now

**Comment by bpohoriletz** [comment]
thank you for the review @composerinteralia, I've made requested changes

**Comment by composerinteralia** [comment]
@bpohoriletz it might be worth rebasing this on `master` so you can get a passing travis build.
@eileencodes I left an initial review, but this still needs review by a committer.

**Comment by bpohoriletz** [comment]
@composerinteralia thank you for the advise, done now

**Comment by oniofchaos** [comment]
@bpohoriletz looks like this just fell through the cracks, can you fix up the merge conflict?

**Comment by bpohoriletz** [comment]
@oniofchaos done

**Comment by oniofchaos** [comment]
@eileencodes would you be able to give this a review? Thanks 💝

**Comment by eileencodes** [comment]
@bpohoriletz thanks for rebasing.

@oniofchaos I know you mean well, but please don't ping reviewers - I already got one email that Bohdan finished rebasing. Thanks for understanding! ❤️

File: activerecord/test/cases/associations/eager_test.rb
Comment (faucct) [inline activerecord/test/cases/associations/eager_test.rb]: These aren't the cases you are fixing: the tests work just fine without the changes.

**Pull request merged**
